src/libraries/System.Private.CoreLib/src/System/Convert.cs

```diff
@@ -2838,5 +2838,97 @@ private static unsafe int FromBase64_ComputeResultLength(char* inputPtr, int inp
             // Done:
             return (usefulInputLength / 4) * 3 + padding;
         }
+
+        /// <summary>
+        /// Converts the specified string, which encodes binary data as hex characters, to an equivalent 8-bit unsigned integer array.
+        /// </summary>
+        /// <param name="s">The string to convert.</param>
+        /// <returns>An array of 8-bit unsigned integers that is equivalent to <paramref name="s"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="s"/> is <code>null</code>.</exception>
+        /// <exception cref="FormatException">The length of <paramref name="s"/>, is not zero or a multiple of 2.</exception>
+        /// <exception cref="FormatException">The format of <paramref name="s"/> is invalid. <paramref name="s"/> contains a non-hex character.</exception>
+        public static byte[] FromHexString(string s)
+        {
+            if (s == null)
+                throw new ArgumentNullException(nameof(s));
+
+            return FromHexString(s.AsSpan());
+        }
+
+        /// <summary>
+        /// Converts the span, which encodes binary data as hex characters, to an equivalent 8-bit unsigned integer array.
+        /// </summary>
+        /// <param name="chars">The span to convert.</param>
+        /// <returns>An array of 8-bit unsigned integers that is equivalent to <paramref name="chars"/>.</returns>
+        /// <exception cref="FormatException">The length of <paramref name="chars"/>, is not zero or a multiple of 2.</exception>
+        /// <exception cref="FormatException">The format of <paramref name="chars"/> is invalid. <paramref name="chars"/> contains a non-hex character.</exception>
+        public static byte[] FromHexString(ReadOnlySpan<char> chars)
+        {
+            if (chars.Length == 0)
+                return Array.Empty<byte>();
+            if (chars.Length % 2 != 0)
```

Comment:
I'm not so sure, but wouldn't a bitwise operation be a little bit faster than computing modulo?

Reply:
If the JIT isn't optimizing this then we should log an issue 😄 

However, this might also be one of the cases where the optimization is only "valid" for unsigned inputs, in which case an analyzer to flag it would also be good.

Reply:
I seem to remember @EgorBo recently teaching the JIT that Array.Length was unsigned. Am I misremembering?

Reply:
The JIT is smart enough to avoid the costly `idiv`.
Casting to `uint` first lets the JIT optimize it even further without loosing (much) readability. 
See here: [https://sharplab.io](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACOSB0ASgK4B2wAllAKb4DCA9lAA4UA21ATgMpcBuFAMbUIAbnTo4AZlwJsdbAG902Vbhl4AbLgAs8gBbVBAa27MwpADLVSAc2D6EACkLUwAEwDypNgE8zFgA8gvpgnAB82CFhEACUKmrKaGop2BQAZthO0ZwQ+NZ2DtgApNhyAIQAvNiY8cmpDQAq+pwMAO5OseL1qgC+Ej3quEjacHp0hiYBVjb2+kgubl4+/uakwaERUZtxCapJDaoZWU7EFOSx2zH5s0WlFdW1e4eqza0dXc/9g88A2gCy1AcDHcAEkWGwnIDgWCIZ5mJQGKQ8gA5Bignznc62WIAXWezE4FD4YGA1GGoz0b3anSU316QA=) 

I'll address that shortly...

Reply:
> I seem to remember @EgorBo recently teaching the JIT that Array.Length was unsigned. Am I misremembering?

@stephentoub He did, here: https://github.com/dotnet/runtime/pull/32368
@EgorBo Could we do the same for `Span<T>` and `ROS<T>`?

Reply:
@stephentoub @tkp1n unfortunately only for Arrays 😢 because we have a special IL opcode for Array.Length - `ldlen` => `GT_ARR_LENGTH` in JIT's IR so it was easy to implement it.
For Spans Length is just a field, e.g.:
```cshap
array.Length % 8
//    *  MOD       int   
//    +--*  ARR_LENGTH int   
//    |  \--*  LCL_VAR   ref    V01 arg1         
//    \--*  CNS_INT   int    8
```
```csharp
mySpan.Length % 8
//   *  MOD       int   
//   +--*  FIELD     int    _length
//   |  \--*  ADDR      byref 
//   |     \--*  LCL_VAR   struct<System.Span`1[Byte], 16> V01 arg1         
//   \--*  CNS_INT   int    8
```
but I'll take a look if there is an easy fix for it